man/newchanges.1
.TH newchanges "1" "December 15, 2023" "" "newchanges manual"

.SH NAME
newchanges - manages changelog file with conventional commit messages

.SH SYNOPSIS
.B newchanges
[-c|--config
.I name
] [-l|--log
.I
file
] [-h|--heading
.I
level
]
.br
[-t|--tag-prefix
.I
prefix
] [-l|--logged-types
.I
types
] [-f|--from
.I
hash
]
.br
[-t|--to
.I
hash
] [-u|--try-unshallow] [-p|--path
.I
path
]
.br
[-r|--repo-url
.I
url
] [-o|--override-version
.I
version
]
.br
[-e|--write-changes
.I
file
] [-w|--write-version
.I
file
]
.br
[-a|--assume-patch] [-0|--bump-major-0] [--pre-release]
.br
[--pre-id
.I
id
] [-d|--dry-run] [-N|--no-failure]
.br
[-i|--print-last] [-q|--quiet] [-v|--verbose]
.br
[-V|--version] [-h|--help]
[
.I command
]
.RE

.SH DESCRIPTION
newchanges creates or updates the changelog file from commit messages formatted according to Conventional Commits.

.SH COMMANDS
.B
.IP "init"
generate a config file with defaults

.SH OPTIONS
.B
.IP "-c|--config name"
file name or path of the config file
.B
.IP "-l|--log file"
file to read from and write to (default: to find)
.B
.IP "-t|--tag-prefix prefix"
expect git tags prefixed (default: "v")
.B
.IP "-h|--heading level"
level of the log entry headings (default: 2)
.B
.IP "-l|--logged-types types"
change types to include in the log
.B
.IP "-f|--from hash"
start at a specific commit (default: last change)
.B
.IP "-t|--to hash"
end at a specific commit (default: HEAD)
.B
.IP "-u|--try-unshallow"
try fetch missing commits and tags if not found
.B
.IP "-p|--path path"
consider only specific path (default: git root)
.B
.IP "-r|--repo-url url"
URL of the git repository (default: from git)
.B
.IP "-o|--override-version version"
set the new version to the specified value
.B
.IP "-e|--write-changes file"
write the new changes to the specified file
.B
.IP "-w|--write-version file"
write the version number to the specified file
.B
.IP "-a|--assume-patch"
assume a patch release for insignificant commits
.B
.IP "-0|--bump-major-0"
bump the major version also if it is 0
.B
.IP "pre-release"
bump the existing pre-release number
.B
.IP "pre-id id"
first pre-release identifier (default: "next")
.B
.IP "-d|--dry-run"
print the new changes on the console only
.B
.IP "-N|--no-failure"
do not fail if the change log was not updated
.B
.IP "-i|--print-last"
print changes for the last version on the console
.B
.IP "-q|--quiet"
omit the summary note on the standard output
.B
.IP "-v|--verbose"
print the new changes on the console too
.B
.IP "-V|--version"
print the version of the executable and exit
.B
.IP "-h|--help"
print the usage information and exit

.RE
Default change types to include in the log: "feat", "fix", "perf". If
the commit message includes the note "BREAKING CHANGE", it will be
included in the log regardless of its type.

.SH EXAMPLES
.B
.IP "newchanges -f v0.1.0 -t v0.2.0"
Update the changelog with changes between two specific tags only.
.B
.IP "newchanges -d"
Generate new changes without modifying the changelog.

.SH BUGS
newchanges's bug tracker is located at <https://github.com/prantlf/v-newchanges/issues>

.SH DOCUMENTATION
See the website for documentation <https://github.com/prantlf/v-newchanges#readme>

.SH AUTHOR
Currently maintained by Ferdinand Prantl <prantlf@gmail.com>


CHANGELOG.md
# Changes

## [0.5.2](https://github.com/prantlf/v-newchanges/compare/v0.5.1...v0.5.2) (2024-04-17)

### Bug Fixes

* Upgrade dependencies ([fa08d9c](https://github.com/prantlf/v-newchanges/commit/fa08d9c3ec93a248550c11948909c9323869e325))

## [0.5.1](https://github.com/prantlf/v-newchanges/compare/v0.5.0...v0.5.1) (2024-03-30)

### Bug Fixes

* Compute pre-releases with prantlf.semvut ([2a91167](https://github.com/prantlf/v-newchanges/commit/2a9116789f936fa0f36277696c3c9e86e6fc2d1a))

## [0.5.0](https://github.com/prantlf/v-newchanges/compare/v0.4.0...v0.5.0) (2024-03-24)

### Features

* Support pre-releases ([c3989cd](https://github.com/prantlf/v-newchanges/commit/c3989cd6a9e6588fa817c1f8f912d8e0e4ad86d8))

## [0.4.0](https://github.com/prantlf/v-newchanges/compare/v0.3.0...v0.4.0) (2023-12-15)

### Features

* Add the man file ([f0c8c32](https://github.com/prantlf/v-newchanges/commit/f0c8c32c871b9a46a1d68c3c8078a4abca9135f1))

## [0.3.0](https://github.com/prantlf/v-newchanges/compare/v0.2.0...v0.3.0) (2023-12-14)

### Features

* Add the linux arm64 support ([0234f3f](https://github.com/prantlf/v-newchanges/commit/0234f3f3f54cf910dc9c9d3013929e4ce308d26c))

## [0.2.0](https://github.com/prantlf/v-newchanges/compare/v0.1.2...v0.2.0) (2023-12-12)

### Features

* Add build for arm64 ([15bd66e](https://github.com/prantlf/v-newchanges/commit/15bd66e8b49fef5ae086b175bc27f49ee89ab896))
* Build for Windows by cross-compiling ([e44880f](https://github.com/prantlf/v-newchanges/commit/e44880f8ee4defe22e23651f8c542d828693b7f9))

## [0.1.2](https://github.com/prantlf/v-newchanges/compare/v0.1.1...v0.1.2) (2023-10-24)

### Bug Fixes

* Build with updated dependencies ([7be45cc](https://github.com/prantlf/v-newchanges/commit/7be45cc933d453c8ab472b8022f3f24d8172c8b7))

## [0.1.1](https://github.com/prantlf/v-newchanges/compare/v0.1.0...v0.1.1) (2023-10-23)

### Bug Fixes

* Do not start the written file log with ./ ([c4304f6](https://github.com/prantlf/v-newchanges/commit/c4304f687d84f7ed56d895b2dacf6950c8c0be69))

## [0.1.0](https://github.com/prantlf/v-newchanges/compare/v0.0.2...v0.1.0) (2023-10-22)

### Features

* Allow extracting the date of the last release from the changelog ([089102d](https://github.com/prantlf/v-newchanges/commit/089102d9e83815d2f806640426369ad368292ab9))
* Try fetching the missing history ([9e7cf16](https://github.com/prantlf/v-newchanges/commit/9e7cf16993725b7994d664d3be4dae9ebc68fd2f))

## [0.0.2](https://github.com/prantlf/v-newchanges/compare/v0.0.1...v0.0.2) (2023-10-22)

### Bug Fixes

* Workaround compiler bug with multiline pattern print on Windows ([267f282](https://github.com/prantlf/v-newchanges/commit/267f2828e3506024b6e715548116be92d2579bd2))

## 0.0.1 (2023-10-21)

Initial release


src/newchanges.v
import time { now }
import os { exists, join_path_single, write_file }
import semver
import prantlf.cli { Cli, run }
import prantlf.debug { rwd }
import prantlf.osutil { exist_in }

const version = '0.5.2'

const usage = 'Updates the changelog file using git log messages.

Usage: newchanges [options] [commands]

Commands:
  init                      generate a config file with defaults

Options:
  -c|--config <name>        file name or path of the config file
  -l|--log <file>           file to read from and write to (default: to find)
  -t|--tag-prefix <prefix>  expect git tags prefixed (default: "v")
  -h|--heading <level>      level of the log entry headings (default: 2)
	-l|--logged-types <types> change types to include in the log
  -f|--from <hash>          start at a specific commit (default: last change)
  -t|--to <hash>            end at a specific commit (default: HEAD)
  -u|--try-unshallow        try fetch missing commits and tags if not found
  -p|--path <path>          consider only specific path (default: git root)
  -r|--repo-url <url>       URL of the git repository (default: from git)
  -o|--override-version <v> set the new version to the specified value
  -e|--write-changes <file> write the new changes to the specified file
  -w|--write-version <file> write the version number to the specified file
  -a|--assume-patch         assume a patch release for insignificant commits
  -0|--bump-major-0         bump the major version also if it is 0
  --pre-release             bump the existing pre-release number
  --pre-id <id>             first pre-release identifier (default: "next")
  -d|--dry-run              print the new changes on the console only
  -N|--no-failure           do not fail if the change log was not updated
  -i|--print-last           print changes for the last version on the console
  -q|--quiet                omit the summary note on the standard output
  -v|--verbose              print the new changes on the console too
  -V|--version              print the version of the executable and exits
  -h|--help                 print the usage information and exits

Default change types to include in the log: "feat", "fix", "perf". If
the commit message includes the note "BREAKING CHANGE", it will be
included in the log regardless of its type.

Examples:
  $ newchanges -f v0.1.0 -t v0.2.0
  $ newchanges -d'

struct Opts {
	quiet            bool
	verbose          bool
	log              string
	tag_prefix       string = 'v'            @[json: 'tag-prefix']
	from             string
	to               string
	try_unshallow    bool              @[json: 'try-unshallow']
	path             string
	repo_url         string            @[json: 'repo-url']
	override_version string            @[json: 'override-version']
	write_changes    string            @[json: 'write-changes']
	write_version    string            @[json: 'write-version']
	assume_patch     bool              @[json: 'assume-patch']
	bump_major_0     bool              @[json: 'bump-major-0']
	pre_release      bool              @[json: 'pre-release']
	pre_id           string = 'next'            @[json: 'pre-id']
	dry_run          bool              @[json: 'dry-run']
	failure          bool = true              @[json: 'failure']
	print_last       bool              @[json: 'print-last']
	commit_sep       string = '----------==========----------'            @[json: 'commit-sep']
	subject_re       string = r'^\s*(?<type>[^: ]+)\s*:\s*(?<description>.+)$'            @[json: 'subject-re']
	body_re          string            @[json: 'body-re']
	footer_re        string            @[json: 'footer-re']
	version_re       string = r'^\s*(?<heading>#+)\s+(?:(?<version>\d+\.\d+\.\d+(?:-[.0-9A-Za-z-]+)?)|(?:\[(?<version>\d+\.\d+\.\d+(?:-[.0-9A-Za-z-]+)?)\])).+\((?<date>[-\d]+)\)\s*$'            @[json: 'version-re']
	prolog           string = '# Changes'
	version_tpl      string = '{heading} [{version}]({repo_url}/compare/{tag_prefix}{prev_version}...{tag_prefix}{version}) ({date})'            @[json: 'version-tpl']
	change_tpl       string  = '#{heading} {title}'            @[json: 'change-tpl']
	commit_tpl       string  = '* {description} ([{short_hash}]({repo_url}/commit/{hash}))'            @[json: 'commit-tpl']
	logged_types     []string = ['feat', 'fix', 'perf']          @[json: 'logged-types'; split]
	type_titles      map[string]string = {
		'feat':            'Features'
		'fix':             'Bug Fixes'
		'perf':            'Performance Improvements'
		'chore':           'Chores'
		'BREAKING_CHANGE': 'BREAKING CHANGES'
	} @[json: 'type-titles']
mut:
	heading int = 2
}

fn main() {
	run(Cli{
		usage: usage
		version: version
		cfg_opt: 'c'
		cfg_gen_arg: 'init'
		cfg_file: '.newchanges'
	}, body)
}

fn body(mut opts Opts, _args []string) ! {
	mut changes_dir := ''
	mut changes_file := ''
	mut changes := []string{}
	mut last_change := -1
	mut last_version := ''
	mut last_date := ''
	changes_file_found := if opts.log.len > 0 {
		changes_file = opts.log
		exists(opts.log)
	} else {
		changes_dir = get_git_root()!
		if changes_dir.len > 0 {
			if changes_name := exist_in(['CHANGELOG.md', 'CHANGES.md'], changes_dir) {
				changes_file = join_path_single(changes_dir, changes_name)
				true
			} else {
				false
			}
		} else {
			false
		}
	}

	if opts.print_last {
		if changes_file_found {
			changes = get_changes(changes_file)!
			first_line, last_line, last_ver := get_last_changes(changes, opts)!
			if first_line < 0 {
				msg := 'no last changes found'
				if !opts.failure {
					println(msg)
					return
				}
				return error(msg)
			}
			contents := if opts.verbose || opts.write_changes.len > 0 {
				changes[first_line..last_line].join('\n')
			} else {
				''
			}
			if opts.verbose {
				print(contents)
			}
			if !opts.quiet {
				if opts.verbose {
					println('\n')
				}
				println('version ${last_ver} and ${last_line - first_line} lines printed')
			}
			if opts.write_changes.len > 0 {
				write_file(opts.write_changes, contents)!
			}
			if opts.write_version.len > 0 {
				write_file(opts.write_version, last_ver)!
			}
			return
		} else {
			return error('change log not found')
		}
	}

	if changes_file_found {
		changes = get_changes(changes_file)!
		mut heading := 0
		last_change, heading, last_version, last_date = analyse_changes(changes, opts)!
		if heading > 0 && heading != opts.heading {
			opts.heading = heading
		}
	}

	commit_log := get_commits(last_version, last_date, opts)!
	mut commits, all_commit_count := parse_commits(commit_log, opts)!
	if commits.len == 0 {
		loc := get_location(last_version, last_date, opts)
		mut msg := if all_commit_count > 0 {
			commit_pl := if all_commit_count > 1 {
				's'
			} else {
				''
			}
			'classified commits from ${all_commit_count} commit${commit_pl}'
		} else {
			'commits'
		}
		msg = 'no ${msg} found ${loc}'
		if !opts.failure {
			println(msg)
			return
		}
		return error(msg)
	}

	next_version, next_date := if opts.override_version.len > 0 {
		semver.from(opts.override_version)!
		date := now().ymmdd()
		opts.override_version, date
	} else {
		ver, date := compute_next_version(commits, last_version, opts)!
		if ver.len == 0 {
			loc := get_location(last_version, last_date, opts)
			commit_pl := if commits.len > 1 {
				's'
			} else {
				''
			}
			msg := 'no significant commits from ${commits.len} classified commit${commit_pl} from ${all_commit_count} total found ${loc}'
			if !opts.failure {
				println(msg)
				return
			}
			return error(msg)
		}
		ver, date
	}

	new_changes := generate_changes(commits, last_version, next_version, next_date, opts)!
	mut prefix := ''
	mut suffix := ''
	mut contents := ''
	if (changes_file.len == 0 && changes_dir.len == 0) || opts.dry_run {
		contents = new_changes.join('\n')
		print(contents)
		if !opts.quiet {
			prefix = '\n'
			suffix = 'printed'
		}
	} else {
		if opts.verbose {
			contents = new_changes.join('\n')
			print(contents)
			if !opts.quiet {
				prefix = '\n'
			}
		}
		if changes_file.len == 0 {
			changes_file = join_path_single(changes_dir, 'CHANGELOG.md')
		}
		if !opts.quiet {
			mut dchanges_file := rwd(changes_file)
			if dchanges_file.starts_with('./') || dchanges_file.starts_with('.\\') {
				dchanges_file = dchanges_file[2..]
			}
			suffix = 'written to "${dchanges_file}"'
		}
		append_changes(changes_file, changes, last_change, new_changes, opts)!
	}
	if !opts.quiet {
		loc := get_location(last_version, last_date, opts)
		commit_pl := if commits.len > 1 {
			's'
		} else {
			''
		}
		println('${prefix}discovered ${commits.len} classified commit${commit_pl} from ${all_commit_count} total ${loc}')
		println('version ${next_version} (${next_date}) and ${new_changes.len} lines ${suffix}')
	}
	if opts.write_changes.len > 0 {
		if contents.len == 0 {
			contents = new_changes.join('\n')
		}
		write_file(opts.write_changes, contents)!
	}
	if opts.write_version.len > 0 {
		write_file(opts.write_version, next_version)!
	}
}

fn get_location(last_version string, last_date string, opts &Opts) string {
	from := if opts.from.len > 0 {
		opts.from
	} else if last_version.len > 0 {
		'${opts.tag_prefix}${last_version} (${last_date})'
	} else {
		'the beginning'
	}
	to := if opts.to.len > 0 {
		' until ${opts.to}'
	} else {
		''
	}
	at := if opts.path.len > 0 {
		' at "${opts.path}"'
	} else {
		''
	}
	return 'since ${from}${to}${at}'
}
